larpandoracontent/LArHelpers/LArHierarchyHelper.cc

```diff
@@ -0,0 +1,1171 @@
+    {
+        try
+        {
+            const MCParticle *pMC{MCParticleHelper::GetMainMCParticle(pCaloHit)};
+            m_mcToHitsMap[pMC].emplace_back(pCaloHit);
+        }
+        catch(const StatusCodeException&)
+        {
+        }
+    }
+
+    if (foldToPrimaries && !foldToLeadingShowers)
+    {
+        MCParticleSet primarySet;
+        m_pNeutrino = LArHierarchyHelper::GetMCPrimaries(mcParticleList, primarySet);
+        MCParticleList primaries(primarySet.begin(), primarySet.end());
+        if (m_recoCriteria.m_removeNeutrons)
+            primaries.erase(std::remove_if(primaries.begin(), primaries.end(), predicate), primaries.end());
+        for (const MCParticle *pPrimary : primaries)
+        {
+            MCParticleList allParticles{pPrimary};
+            if (!m_recoCriteria.m_removeNeutrons)
+            {
+                LArMCParticleHelper::GetAllDescendentMCParticles(pPrimary, allParticles);
+            }
+            else
+            {   // Collect track-like and shower-like particles together, but throw out neutrons and descendents
+                MCParticleList neutrons;
+                LArMCParticleHelper::GetAllDescendentMCParticles(pPrimary, allParticles, allParticles, neutrons);
+            }
+            CaloHitList allHits;
+            for (const MCParticle *pMC : allParticles)
+            {
+                // Not all MC particles will have hits
+                if (m_mcToHitsMap.find(pMC) != m_mcToHitsMap.end())
+                {
+                    const CaloHitList &caloHits(m_mcToHitsMap.at(pMC));
+                    allHits.insert(allHits.begin(), caloHits.begin(), caloHits.end());
+                }
+            }
+            m_rootNodes.emplace_back(new Node(*this, allParticles, allHits));
+        }
+    }
+    else if (foldToPrimaries && foldToLeadingShowers)
+    {
+        MCParticleSet primarySet;
+        m_pNeutrino = LArHierarchyHelper::GetMCPrimaries(mcParticleList, primarySet);
+        MCParticleList primaries(primarySet.begin(), primarySet.end());
+        if (m_recoCriteria.m_removeNeutrons)
+            primaries.erase(std::remove_if(primaries.begin(), primaries.end(), predicate), primaries.end());
+        for (const MCParticle *pPrimary : primaries)
+        {
+            MCParticleList allParticles{pPrimary}, showerParticles, neutrons;
+            int pdg{std::abs(pPrimary->GetParticleId())};
+            const bool isShower{pdg == E_MINUS || pdg == PHOTON};
+            const bool isNeutron{pdg == NEUTRON};
+            if (isShower || isNeutron)
+            {
+                if (!m_recoCriteria.m_removeNeutrons)
+                {
+                    LArMCParticleHelper::GetAllDescendentMCParticles(pPrimary, allParticles);
+                }
+                else
+                {   // Throw away neutrons
+                    MCParticleList dummy;
+                    LArMCParticleHelper::GetAllDescendentMCParticles(pPrimary, allParticles, allParticles, dummy);
```

Comment:
How come the neutron list is sometimes called "dummy", sometimes "neutron", if the contents are discarded anyway? Could the same list be reused?